File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

int maxValueOfArray(int *arr, int len){
    int m = -1;
    
    for(int i = 0; i < len; i++){
        if(arr[i] > m)
            m = arr[i];
    }
    return m;
}

int pairSum(struct ListNode* head)
{
    struct ListNode *tmp = head;
    int i = 0,len = 0,max;
    
    while(tmp){
        len++;
        tmp = tmp->next;
    }
    int *arr;
    arr = (int *)calloc(len/2, sizeof(int));
    
    tmp = head;
    while(tmp){
        if(i < len/2)
            arr[i] = tmp->val;
        else
            arr[len-1-i] += tmp->val; 
        tmp = tmp->next;
        i++;
    }
    max = maxValueOfArray(arr,len/2);
    return max;
}

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        Stack<Integer> stack = new Stack<>();
        ListNode tmp = head;
        int max = 0;
        int len = 0;
        while(tmp != null){
            len++;
            tmp = tmp.next;
        }
        tmp = head;
        for(int i = 0; i < len/2; i++){     //first half
            stack.push(tmp.val);
            tmp = tmp.next;
        }
        while(!stack.empty()){              //second half
            if(stack.peek() + tmp.val > max)
                max = stack.peek() + tmp.val;
            stack.pop();
            tmp = tmp.next;
        }
        return max;
    }
}